### ISSUE TITLE
Bug: v6 `typeMatchesSpecifier` `PackageSpecifier` should match custom type roots

### ISSUE THREAD

**Issue by RebeccaStevens** [typescript-eslint/typescript-eslint#6868]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```typescript
typeMatchesSpecifier(
  type,
  {
    from: "package",
    package: "foo"
    name: "Foo",
  },
  parserServices.program)
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "typeRoots": ["./typings"],
    "baseUrl": "./",
    "paths": {
      "*": ["typings/*"]
    }
  }
}
```

### Expected Result

Type declarations in `typings` should be treated as external.

### Actual Result

Type declarations in `typings` are treated as local files.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.0.0-alpha.121` |
| `@typescript-eslint/parser`            | `6.0.0-alpha.121` |
| `@typescript-eslint/scope-manager`     | `6.0.0-alpha.121` |
| `@typescript-eslint/typescript-estree` | `6.0.0-alpha.121` |
| `@typescript-eslint/type-utils`        | `6.0.0-alpha.121` |
| `@typescript-eslint/utils`             | `6.0.0-alpha.121` |
| `TypeScript`                           | `5.0.3` |
| `ESLint`                               | `8.36.0` |
| `node`                                 | `18.15.0` |
